File: animations/strobe.py
from time import sleep, time_ns
from rpi_ws281x import Color as StripColor

from __init__ import ANIMATION_STEPS
from animations.animation import Animation


class Strobe(Animation):
    def select_range(self, index: int, colors: list) -> list:
        animation_range = []

        match self.direction:
            case 0:  # Normal
                animation_range.append(range(int(self.start_led + 1 / len(colors)
                                                 * (self.end_led - self.start_led)
                                                 * index),
                                             int(self.end_led + 1
                                                 - 1 / len(colors)
                                                 * (self.end_led - self.start_led)
                                                 * (len(colors) - index - 1))))

            case 1:  # Reverse
                animation_range.append(range(int(self.start_led + 1 / len(colors)
                                                 * (self.end_led - self.start_led)
                                                 * (len(colors) - index - 1)),
                                             int(self.end_led + 1
                                                 - 1 / len(colors)
                                                 * (self.end_led - self.start_led)
                                                 * index)))

            case 2:  # Innter to outer
                animation_range.append(range(int(self.start_led + 1 / len(colors)
                                                 * ((self.end_led - self.start_led) / 2)
                                                 * index),
                                             int(self.end_led + 1
                                                 - 1 / len(colors)
                                                 * ((self.end_led - self.start_led) / 2)
                                                 * (len(colors) - index - 1)
                                                 - (self.end_led - self.start_led) / 2)))

                animation_range.append(range(int(self.start_led
                                                 + 1 / len(colors)
                                                 * ((self.end_led - self.start_led) / 2)
                                                 * (len(colors) - index - 1)
                                                 + (self.end_led - self.start_led) / 2),
                                             int(self.end_led + 1
                                                 - 1 / len(colors)
                                                 * ((self.end_led - self.start_led) / 2)
                                                 * index)))

            case 3:  # Outer to inner
                animation_range.append(range(int(self.start_led + 1 / len(colors)
                                                 * ((self.end_led - self.start_led) / 2)
                                                 * (len(colors) - index - 1)),
                                             int(self.end_led + 1
                                                 - 1 / len(colors)
                                                 * ((self.end_led - self.start_led) / 2)
                                                 * index
                                                 - (self.end_led - self.start_led) / 2)))

                animation_range.append(range(int(self.start_led
                                                 + 1 / len(colors)
                                                 * ((self.end_led - self.start_led) / 2)
                                                 * index
                                                 + (self.end_led - self.start_led) / 2),
                                             int(self.end_led + 1
                                                 - 1 / len(colors)
                                                 * ((self.end_led - self.start_led) / 2)
                                                 * (len(colors) - index - 1))))

        return animation_range
    
    def animate(self, beat: int, step: int):
        colors = self.select_colors(beat)
        third = round(ANIMATION_STEPS * .33)
        half = round(ANIMATION_STEPS * .5)
        two_third = round(ANIMATION_STEPS * .66)
        half_third = round(ANIMATION_STEPS * .125)

        # Direction is used to determine how often a flash occurs
        if self.direction == 0 or (self.direction == 1 and beat % 2 == 0) or \
           (self.direction == 2 and beat % 4 == 0) or (self.direction == 3 and beat % 8 == 0):
            if self.variation == 0:  # Flash on the beat and turn off after 1/3 of the beat
                if step == 0:
                    # Set the color of the strip to the beat color
                    for index, color in enumerate(colors):
                        animation_range = self.select_range(index, colors)

                        if len(animation_range) == 0:
                            break

                        for i in animation_range:
                            for j in i:
                                self.set_pixel_color(j, color)
                
                elif step == third:
                    # Turn off the strip
                    for i in range(self.start_led, self.end_led + 1):
                        self.set_pixel_color(i, StripColor(0, 0, 0))
            
            elif self.variation == 1:  # Flash twice per beat and each flash turns off after 1/6 of the beat
                if step == 0 or step == half:
                    # Set the color of the strip to the beat color
                    for index, color in enumerate(colors):
                        animation_range = self.select_range(index, colors)

                        if len(animation_range) == 0:
                            break

                        for i in animation_range:
                            for j in i:
                                self.set_pixel_color(j, color)
                
                elif step == half_third or step == round(ANIMATION_STEPS * .66):
                    # Turn off the strip
                    for i in range(self.start_led, self.end_led + 1):
                        self.set_pixel_color(i, StripColor(0, 0, 0))

            elif self.variation == 2:  # Flash three times per beat and each flash turns off after 1/8 of the beat
                if step == 0 or step == third or step == two_third:
                    # Set the color of the strip to the beat color
                    for index, color in enumerate(colors):
                        animation_range = self.select_range(index, colors)

                        if len(animation_range) == 0:
                            break

                        for i in animation_range:
                            for j in i:
                                self.set_pixel_color(j, color)

                elif step == half_third or step == third + half_third or step == two_third + half_third:
                    # Turn off the strip
                    for i in range(self.start_led, self.end_led + 1):
                        self.set_pixel_color(i, StripColor(0, 0, 0))

File: src/animation/Strobe.cpp
#include "Strobe.hpp"
#include "../Color.hpp"
#include <cmath>
#include <iostream>

using namespace std;

std::vector<std::vector<int>> Strobe::select_range(int index, const std::vector<Color>& colors) {
    std::vector<std::vector<int>> animation_range;
    int len = colors.size();
    int total_leds = end_led - start_led + 1;

    auto get_range = [&](float start_ratio, float end_ratio) -> std::vector<int> {
        int start = static_cast<int>(start_led + total_leds * start_ratio);
        int end = static_cast<int>(start_led + total_leds * end_ratio);
        std::vector<int> range;
        for (int i = start; i < end; ++i) {
            range.push_back(i);
        }
        return range;
    };

    switch (direction) {
        case 0:  // Normal
            animation_range.push_back(get_range(
                1.0f / len * index,
                1.0f - 1.0f / len * (len - index - 1)));
            break;
        case 1:  // Reverse
            animation_range.push_back(get_range(
                1.0f / len * (len - index - 1),
                1.0f - 1.0f / len * index));
            break;
        case 2:  // Inner to outer
            animation_range.push_back(get_range(
                1.0f / len * index / 2.0f,
                1.0f - 1.0f / len * (len - index - 1) / 2.0f - 0.5f));
            animation_range.push_back(get_range(
                0.5f + 1.0f / len * (len - index - 1) / 2.0f,
                1.0f - 1.0f / len * index / 2.0f));
            break;
        case 3:  // Outer to inner
            animation_range.push_back(get_range(
                1.0f / len * (len - index - 1) / 2.0f,
                1.0f - 1.0f / len * index / 2.0f - 0.5f));
            animation_range.push_back(get_range(
                0.5f + 1.0f / len * index / 2.0f,
                1.0f - 1.0f / len * (len - index - 1) / 2.0f));
            break;
    }

    return animation_range;
}

void Strobe::animate(int beat, int step) {
    std::vector<Color> colors = select_colors(beat);
    int third = static_cast<int>(round(ANIMATION_STEPS * 0.33));
    int half = static_cast<int>(round(ANIMATION_STEPS * 0.5));
    int two_third = static_cast<int>(round(ANIMATION_STEPS * 0.66));
    int half_third = static_cast<int>(round(ANIMATION_STEPS * 0.125));

    bool should_flash = direction == 0 || (direction == 1 && beat % 2 == 0) ||
                        (direction == 2 && beat % 4 == 0) || (direction == 3 && beat % 8 == 0);

    if (!should_flash) return;

    auto turn_on = [&]() {
        for (size_t index = 0; index < colors.size(); ++index) {
            auto animation_ranges = select_range(static_cast<int>(index), colors);
            for (const auto& range : animation_ranges) {
                for (int led : range) {
                    set_pixel_color(led, colors[index].red, colors[index].green, colors[index].blue);
                }
            }
        }
    };

    auto turn_off = [&]() {
        for (int i = start_led; i <= end_led; i++) {
            Color c = black();
            set_pixel_color(i, 0, 0, 0);
        }
    };

    switch (variation) {
        case 0:
            if (step == 0) {
                turn_on();
            } else if (step == third) {
                turn_off();
            }
            break;
        case 1:
            if (step == 0 || step == half) {
                turn_on();
            } else if (step == half_third || step == two_third) {
                turn_off();
            }
            break;
        case 2:
            if (step == 0 || step == third || step == two_third) {
                turn_on();
            } else if (step == half_third || step == third + half_third || step == two_third + half_third) {
                turn_off();
            }
            break;
    }
}
